### ISSUE TITLE
get from local storage at application start

### ISSUE THREAD

**Issue by JR-Utily** [yaacovCR/angular-stored-object#2]
Hello,

This project is very promising, but I think it miss one main use case of local storage : to be able to get back information from previous session when starting the application.

$create should not check the key is not yet used, and instanciate the value with what is found in the local storage, when the strategy used is localStorage

Or at least there should be a $get(strategey) method doing something similar.

**Comment by yaacovCR** [comment]
From the documentation (http://yaacovcr.github.io/angular-stored-object/docs/#/api/yaacovCR.storedObject.type:StoredObject), bolding the relevant:

> Properties of the object can be set just as with any other Javascript object. **If an object with the same key has been saved to storage by this or any other tab, the new object will be loaded from storage and will contain the properties that were initially set;** otherwise, the object will have no properties besides the built-in methods.

Basically, I tried to model behavior on $resource (https://docs.angularjs.org/api/ngResource/service/$resource), but I tried to simplify things a bit by automatically loading the instance from storage if possible, obviating the need for $get. Let me know if that does not make sense? If it's not working for you, i.e., you have found a bug, let me know, otherwise, I will mark this issue as closed...

To put it another way, the factory should already be doing what you want it to do. Check out how simple it is to set up (https://github.com/yaacovCR/glass/blob/master/src/app/components/auth/session.service.js)

```
  angular
    .module('glass.auth')
    .factory('session', session);

  function session(ycr$StoredObject) {
    return new ycr$StoredObject('session');
  }
```
And how whether the user is logged can be checked just by whether the token property from the session service is defined (https://github.com/yaacovCR/glass/blob/master/src/app/components/auth/auth.service.js):

```
      function isLoggedIn() {
        return !!session.token;
      }
```
And how the user can be redirected to the correct state with help from your router of choice, in this case angular-ui (https://github.com/yaacovCR/glass/blob/master/src/app/components/auth/auth.service.js, comments added):

```
    this.$get = function($rootScope, $timeout, $state, $mdDialog, ServerSession, ServerUser, session) {      
      $rootScope.$on('$stateChangeStart', onStateChangeStart);
      $rootScope.$on('storedObject:session:externalChange', onSessionChange);
```

<snip>

```
      function onStateChangeStart(event, toState, toParams) {
        $timeout(function() {
          if (toState.name === _loginState && isLoggedIn()) {
          //no need to log in if already logged in,
          //and if for some reason has been redirected to login
          //we can go send the user where they wanted to go
            event.preventDefault();
            followRedirect();
          }
          if (toState.name !== _loginState && toState.data && toState.data.authenticate && !isLoggedIn()) {
          //if this state required authentication, and we are not logged in,
          //send user to login, and save requested state for later redirection
            event.preventDefault();
            registerRedirect(toState, toParams);
            $state.transitionTo(_loginState);
          }
        });
      }
```

**Comment by yaacovCR** [comment]
Here is the code where the automatically loading is done (https://github.com/yaacovCR/angular-stored-object/blob/master/src/stored-object.js)

```
      _init();
      
      function _init() {
        _key = 'storedObject:' + _key;
        _addEventHandler();
        if ($window.sessionStorage.getItem(_key) != null || $window.localStorage.getItem(_key) != null) {
          var storageType = ($window.sessionStorage.getItem(_key) != null) ? 'sessionStorage' : 'localStorage';
          _loadIntoMemory(angular.fromJson($window[storageType].getItem(_key)));
        } else {
          _triggerEvent('requested', Date.now());
        }
      }
```

**Comment by JR-Utily** [comment]
Sorry!
I was searching the code for localStorage.getItem, and as I didn't find it, I assumed it was missing. My bad.
Thanks for your quick reply !

**Comment by yaacovCR** [comment]
No problem. Hope you find it helpful.

**Comment by JR-Utily** [comment]
I keep talking here, but I can open a new issue if needed
So in fact, knowing it gets the data from memory at initialization, there is no way to know if some data was really read or if it is a new object.
I wanted to make a systematic $create() in the factory to allow me to just call $update() in the main code, but then it throw error at successive runs.

my workaround for now is to encapsulate the $create in a try/catch, but it would be better, i think, to have a $isInit() method telling if we should call create or update for next allocation.

**Comment by yaacovCR** [comment]
See above code. In my use, I check to see whether a predefined property is
defined ('token') and only call $create if undefined. I see you may need an
$isInit method if you have loaded an object from local storage with no
reliable property to check, but is that really what you are doing? I am
curious what your use case is. If necessary, $isInit may be useful, and can
be added, but it may be unnecessary... Can you describe your use case a bit
or include some code?

On Fri, Jan 27, 2017, 10:17 AM JR Utily <notifications@github.com> wrote:

I keep talking here, but I can open a new issue if needed
So in fact, knowing it gets the data from memory at initialization, there
is no way to know if some data was really read or if it is a new object.
I wanted to make a systematic $create() in the factory to allow me to just
call $update() in the main code, but then it throw error at successive runs.

my workaround for now is to encapsulate the $create in a try/catch, but it
would be better, i think, to have a $isInit() method telling if we should
call create or update for next allocation.

—
You are receiving this because you modified the open/close state.


Reply to this email directly, view it on GitHub
<https://github.com/yaacovCR/angular-stored-object/issues/2#issuecomment-275689071>,
or mute the thread

**Comment by JR-Utily** [comment]
Here is what I currently do (in typescript, but this should change nothing):

```

/** @ngInject */
  function StoredCompany(ycr$StoredObject: any): StoredObject {
    let storage = new ycr$StoredObject(COMPANY_KEY);
    try {
      storage.$create(LOCAL_STORAGE_STRATEGY);
    } catch(e){};
    return storage;
  }

angular.module('aaa').factory('storedCompany', StoredCompany);

// and in my service ...

 saveProfile = (profile: IProfile) => {
      if (profile.company) {
        this.storedCompany.data= profile.company;
        this.storedCompany.$update();
      }
}

```

I wanted to keep the service with just one method, I prefer it not to bother if this is a create or an update, so I made the creation at first in the factory.
I could test if `this.storedCompany.data` is already defined before allocating it too, and then routing to one method or the other... 

It seems to me that checking the state each time you use the variable is not very maintainable, it is easy to have someone in the team forgeting to do it. On the other side, I like the idea to have an object I can trust when I get it from injection.

**Comment by yaacovCR** [comment]
I agree that you should be doing the check only once per service, I would
place it in the run block. Create and update differ in that create requires
a storage strategy. In theory, a combined method could be used that creates
or updates as necessary and use a default storage strategy, that is perhaps
specified when the service is created. I think this is a reasonable feature
request, especially useful when the storage strategy is not likely to
change. I suggest you start a new issue as a feature request and I will
work on it. Or, you could fork and submit a pull request.
